### PR TITLE
Add MinGW support (gcc on Windows)

### PR DESCRIPTION
**Pull request by KlamerS**
Apply patches as send earlier in issue #999 to code.

In addition:
* Make MinGW specifics in the Makefile conditional to OS Windows_NT
* Make multiple definition errors go away by adding -Wl,--allow-multiple-definition.

### CODE REVIEW

**Comment by abadams** [comment]
I pulled this into a branch. I have figured out a few things to work around the math errors (mostly by providing the missing functions in a mingw-specific runtime module). I'll get all the tests working and merge it into master.

**Comment by zvookin** [comment]
Thank you for submitting as a PR.

**Comment by abadams** [comment]
Yeah, thanks for this work. It makes it more pleasant for me to hack on Halide at home - I only have a Windows box but I dislike visual studio.

File: test/correctness/cross_compilation.cpp [hunk 1]
```diff
@@ -37,7 +37,7 @@ int main(int argc, char **argv) {
 
         #ifndef _MSC_VER
         std::string object_name = "test_object_" + t + ".o";
```
Comment (zvookin) [inline test/correctness/cross_compilation.cpp]: Can we fix this to not form the string ".obj" via concatenation? It is really nice to be able to grep code for "obj" and find all occurrences and in this case it seems trivial to just append ".o" or ".obj" based on the platform. (Yes, I am aware that the problem is not introduced by this PR, but it seems easy to fix here.)

Comment (abadams) [inline test/correctness/cross_compilation.cpp]: Yeah, that code is a little too clever. Fixed in trunk and merged into the
branch.

**Comment by abadams** [comment]
This continues in #1009

**Pull request closed without merging**